TESTING.md

```diff
@@ -0,0 +1,115 @@
+
+# Test categories
+
+## Unit tests
+
+Unit tests are used to test particular code snippets in isolation. They are your best ally when looking for quick feedback loops in a particular function. 
+
+Imagine you're working on a PR that adds a new field to the Prometheus CRD and you want to test if your change is reflected to the statefulset. Instead of creating a full Kubernetes cluster, installing all the CRDs, running the Prometheus-Operator, deploying a Prometheus resource and finally check if your change made it to the live object, you could simply write or extend a unit test for the statefulset generation.
+
+Here is an example test that checks if labels are present in the statefulset.
+
+https://github.com/prometheus-operator/prometheus-operator/blob/aeceb0b4fadc8307a44dc55afdceca0bea50bbb0/pkg/prometheus/server/statefulset_test.go#L140-L167
+
+Unit tests can be run with:
+
+```
+make test-unit
+```
+
+They can also be run for particular packages:
+
+```
+go test ./pkg/prometheus/server
+```
+
+Or even particular functions:
+
+```
+go test -run ^TestPodLabelsAnnotations$ ./pkg/prometheus/server
+```
+
+### Testing multiline string comparison - Golden files
+
+Golden files are plaintext documents designed to facilitate the validation of lengthy strings. They come in handy when, for instance, you need to test a Prometheus configuration that's generated using Go structures. You can marshal this configuration into YAML and then compare it against a static reference to ensure a match. Golden files offer an elegant solution to this challenge, sparing you the need to hard-code the static configuration directly into your test code.
+
+In the example below, we're generating the Prometheus configuration (which can easily have 100+ lines for each individual test) and comparing it against a golden file:
+
+https://github.com/prometheus-operator/prometheus-operator/blob/aeceb0b4fadc8307a44dc55afdceca0bea50bbb0/pkg/prometheus/promcfg_test.go#L102-L277
+
+If not for golden files, the test above, instead of ~150 lines, would easily require around ~1000 lines. The usage of golden files help us maintain test suites with several multiline strings comparison without sacrifing test readability.
+
+### Updating Golden Files
+
+There are contributions, e.g. adding a new required field to an existing configuration, that requires us to update several golden files at once. This can easily be done with the command below:
+
+```
+make test-unit-update-golden
+```
+
+## End-to-end tests
+
+Sometimes, running tests in isolation is not enough and we really want test the behavior of Prometheus-Operator when running in a working Kubernetes cluster. For those occasions, end-to-end tests are our choice.
+
+To run e2e-tests, first start a Kubernetes cluter. We recommend using [MiniKube](https://minikube.sigs.k8s.io/docs/start/) or [KinD](https://kind.sigs.k8s.io/) for development purposes since they are light weight distributions of kubernetes and can run on small notebooks.
```

Comment:
(nit) I would put Kind first since this is what the CI uses.
```suggestion
To run e2e-tests locally, first start a Kubernetes cluster. We recommend [KinD](https://kind.sigs.k8s.io/) because it is lightweight (it can run on small notebooks) and this is what the project's CI uses.  MiniKube](https://minikube.sigs.k8s.io/docs/start/) is also another option.
```